contributors/design-proposals/network/support_traffic_shaping_for_kubelet_cni.md

```diff
@@ -0,0 +1,72 @@
+# Support traffic shaping for CNI network plugin
+
+Authors: @m1093782566
+
+## Motivation and background
+
+Currently the kubenet code supports applying basic traffic shaping during pod setup. This will happen if bandwidth-related annotations have been added to the pod's metadata, for example:
+
+```json
+{
+    "kind": "Pod",
+    "metadata": {
+        "name": "iperf-slow",
+        "annotations": {
+            "kubernetes.io/ingress-bandwidth": "10M",
+            "kubernetes.io/egress-bandwidth": "10M"
+        }
+    }
+}
+```
+
+Our current implementation uses the `linux tc` to add an download(ingress) and upload(egress) rate limiter using 1 root `qdisc`, 2 `class `(one for ingress and one for egress) and 2 `filter`(one for ingress and one for egress attached to the ingress and egress classes respectively).
+
+Kubelet CNI code doesn't support it yet. However, there is enough functionality for us to add a simple CNI plugin to replicate the behavior we have today in kubenet if we feel this is an important feature. In fact, there is already a [PR](https://github.com/containernetworking/plugins/pull/96) in flight to create a traffic shaping plugin for CNI.
+
+## Goal
+
+Support traffic shaping for CNI network plugin in Kubernetes.
+
+## Non-goal
+
+CNI plugins to implement this sort of traffic shaping guarantee.
+
+## Proposal
+
+This proposal is similar to how host port remapping is done. Once cni plugins support traffic shaping, and Kubelet start up with network plugin == cni, it will work with traffic shaping. traffic shaping will work with ptp and bridge plugins.
+
+### Pod Setup
+
+When we create a pod with bandwidth configuration in its metadata, for example, 
+
+```json
+{
+    "kind": "Pod",
+    "metadata": {
+        "name": "iperf-slow",
+        "annotations": {
+            "kubernetes.io/ingress-bandwidth": "10M",
+            "kubernetes.io/egress-bandwidth": "10M"
+        }
+    }
+}
+```
+
+Kubelet will parse the ingress and egress bandwidth values and transform them to Kb because both `ingressRate` and `egressRate` in cni bandwidth plugin are in Kbps. Eventually, kubelet will bulid the runtime config which like the following and call cni plugin `AddNetworkList` API.
+
+```json
```

Comment:
I've asked on the upstream plugin regarding support for runtimeConfig / capabilities similar to how the portmap plugin works today: https://github.com/containernetworking/plugins/pull/96#issuecomment-370952052

I think we want to agree a capability and use that to negotiate whether or not shaping will be done.

So a user would add something like this to their CNI config list if they wanted to enable traffic shaping via the plugin:

```json
{
  "type": "bandwidth",
  "capabilities": {"trafficShaping": true}
}
```

The kubelet would then populate the `runtimeConfig` section of the config when calling the `bandwidth` plugin:

```json
{
  "type": "bandwidth",
  "runtimeConfig": {
    "trafficShaping": {
      "ingressRate": "X",
      "egressRate": "Y"
    }
  }
}
```